lib/speech/speech_recognizer.ts

```diff
@@ -0,0 +1,233 @@
+        // this is important, as the protocol does not allow chunks larger than 8192 bytes
+        super({ highWaterMark: 8192 });
+
+        this._stream = stream;
+
+        this._endDetected = false;
+        this._ended = false;
+        this._started = false;
+
+        //this._debugFile = fs.createWriteStream('out_' + process.pid + '_' + (i++) + '.wav');
+        this._vad = vad;
+        this.consecutiveSilence = 0;
+
+        // all chunks received before the connection is ready are buffered here
+        this._bufferedMessages = [];
+
+        this.write(initialBuffer);
+        this._stream.pipe(this);
+        this._piped = true;
+
+        this.on('error', () => {
+            if (!this._piped)
+                return;
+            this._piped = false;
+            this._stream.unpipe(this);
+        });
+    }
+
+    start(connection : WebSocket) {
+        this._started = true;
+        this._connection = connection;
+
+        this._listener = this._handleMessage.bind(this);
+        this._connection.on('message', this._listener);
+
+        this._endTimeout = setTimeout(() => this.end(), 150000);
+        for (let message of this._bufferedMessages)
+            this._connection.send(message, { binary: true });
+        this._bufferedMessages = [];
+    }
+
+    _finish(callback : () => void) {
+        // readable stream ended
+        this._piped = false;
+        this.end();
+        callback();
+    }
+
+    end() {
+        if (this._piped) {
+            this._piped = false;
+            this._stream.unpipe(this);
+        }
+
+        if (this._ended)
+            return;
+
+        // end() before start() indicates an error connecting to the server (e.g.
+        // access token error)
+        if (!this._started) {
+            this._ended = true;
+            return;
+        }
+
+        if (this._endTimeout) clearTimeout(this._endTimeout);
+
+        this._ended = true;
+        //this._debugFile.end();
+    }
+
+    _handleMessage(msg : Buffer) {
+        //console.log('Received message');
+        //console.log(msg);
+        let json;
+        try {
+            json = JSON.parse(msg.toString());
+        } catch(e) {
+            this.emit('error', e);
+            this.end();
+            return;
+        }
+
+        if (json.status === 400 && json.code === "E_NO_MATCH")
+            this.emit('done', "NoMatch");
+        else if (json.result === 'ok')
+            this.emit('done', "Success", json.text);
+        else
+            this.emit('error', msg);
+        this.end();
+    }
+
+    _write(chunk : Buffer, encoding : BufferEncoding, callback : (err ?: Error) => void) {
+        if (this._ended || this._endDetected) {
+            callback();
+            return;
+        }
+        //console.log('Sending chunk of length ' + chunk.length);
+        if (this._connection && this._connection.readyState === 1) {
+            // OPEN
+            if (this._vad && chunk.length === 320) {
```

Comment:
Why are we checking for an exact length here? node will merge and split buffers in the stream code